packager/packager.cc

```diff
@@ -275,6 +275,73 @@ class PackageInfoImpl final : public core::packages::PackageInfo {
     explicit PackageInfoImpl(const PackageInfoImpl &) = default;
     PackageInfoImpl &operator=(const PackageInfoImpl &) = delete;
 
+    // What order should these packages be in the import list?
+    // Returns -1 if a should come before b, 0 if they are equivalent, and 1 if a should come after b.
+    // TODO(neil): sort by alphabetical order too
+    int orderByStrictness(const core::packages::PackageDB &packageDB, const PackageInfoImpl &a,
+                          const PackageInfoImpl &b) const {
+        if (!strictDependenciesLevel.has_value() || !a.strictDependenciesLevel.has_value() ||
+            !b.strictDependenciesLevel.has_value() || !a.layer.has_value() || !b.layer.has_value()) {
+            return 0;
+        }
+        // Layering violations always come first
+        if (causesLayeringViolation(packageDB, a)) {
+            if (causesLayeringViolation(packageDB, b)) {
+                return 0;
+            } else {
+                return -1;
+            }
+        }
+
+        auto aStrictDependenciesLevel = a.strictDependenciesLevel.value().first;
+        auto bStrictDependenciesLevel = b.strictDependenciesLevel.value().first;
+        switch (strictDependenciesLevel.value().first) {
```

Comment:
Does this get simpler if you map levels to integers and compare the integers?

Reply:
Or maybe this is complicated because of something weird about how gen-packages sorts things differently in packages of a given strictness level?

Reply:
Yup, it's a bit more complex because it's a function of the strictness level, and it's not a straightforward "false < layered < ..." ordering. I added a comment explaining the ordering.